### PR TITLE
Logging improvements

### LINKED ISSUE #14: add support to report size of compiled sketch
https://arduino.stackexchange.com/questions/8058/how-can-arduino-source-code-be-tested-automatically-continuous-integration

### PR DESCRIPTION
**Pull request by ianfixes**
## Highlights from `CHANGELOG.md`

* Using new logging class to remove arbitrary whitespace placement in favor of explicit functionality for adjusting indentation levels.
* See CHANGELOG.md for more

### TODO:
* #321 

## Issues Fixed

* Fixes #14

### CODE REVIEW

File: exe/arduino_ci.rb [hunk 1]
```diff
@@ -424,16 +388,16 @@ def perform_unit_tests(cpp_library, file_config)
   platforms = choose_platform_set(config, "unittest", config.platforms_to_unittest, cpp_library.library_properties)
   if platforms.empty?
     explain_and_exercise_envvar(VAR_EXPECT_UNITTESTS, "unit tests", "platforms and architectures") do
-      puts "    Configured platforms: #{config.platforms_to_unittest}"
-      puts "    Configuration is default: #{config.is_default}"
+      @log.iputs "    Configured platforms: #{config.platforms_to_unittest}"
```
Comment (hlovdal) [inline exe/arduino_ci.rb:412]: I guess these could be wrapped with `@log.indent`.

File: exe/arduino_ci.rb [hunk 2]
```diff
@@ -520,83 +484,64 @@ def perform_example_compilation_tests(cpp_library, config)
...
-          puts @backend.last_err
+      compiled_ok = @log.attempt("Compiling #{example_name} for #{board}") do
+        @backend.compile_sketch(example_path, board)
+      end
+
+      # decode the JSON output of the compiler a little bit
+      unless compiled_ok
+        @log.inform_multiline("Compilation failure details") do
+          begin
+            # parse the JSON, and print out only the nonempty keys. indent them with 4 spaces in their own labelled sections
```
Comment (hlovdal) [inline exe/arduino_ci.rb:524]: The comment mentions added indentation, but I think it is missing from the code.

File: exe/arduino_ci.rb (hunk 1 above)
Comment (ianfixes) [reply exe/arduino_ci.rb:412]: Good catch, thanks!  Refactoring whitespace is surprisingly difficult 😅

File: exe/arduino_ci.rb (hunk 2 above)
Comment (ianfixes) [reply exe/arduino_ci.rb:524]: `inform_multiline` implicitly indents, so the full magic there is `@log.inform_multiline(k) { @log.iputs(val) } unless val.strip.empty?`

**Comment by hlovdal** [comment]
The failing builds are due to stale old references to plain `inform_multiline` as opposed to new `@log.inform_multiline`.

```text
exe/arduino_ci.rb
362:    return inform_multiline("Wildcard architecture in library.properties, using configured platforms") do
373:    return inform_multiline("Default config, platforms matching architectures in library.properties") do
382:  inform_multiline("Configured platforms that match architectures in library.properties") do
```

**Comment by ianfixes** [comment]
Yup, I'm just making a lot of local edits and pushing up stuff to my branch that (for now) I expect to have failures.  Lots of areas being cleaned up!

**Pull request merged**